compiler/src/main/java/com/squareup/anvil/compiler/codegen/incremental/GeneratedFileCache.kt

```diff
@@ -0,0 +1,238 @@
+    }
+  }
+
+  fun addSourceFile(sourceFile: RelativeFile) {
+    addMd5(sourceFile, overwrite = true)
+  }
+
+  private fun addMd5(sourceFile: RelativeFile, overwrite: Boolean) {
+    if (overwrite) {
+      filesToMd5[sourceFile] = sourceFile.md5()
+    } else {
+      filesToMd5.computeIfAbsent(sourceFile) { sourceFile.md5() }
+    }
+  }
+
+  fun addGeneratedFile(generated: GeneratedFileWithSources) {
+
+    val generatedAbsolute = AbsoluteFile(generated.file)
+    val generatedRelative = generatedAbsolute.relativeTo(projectDir)
+
+    generatedToContent[generatedRelative] = generated.content
+
+    addMd5(generatedRelative, overwrite = true)
+
+    filesToMd5.computeIfAbsent(generatedRelative) { generatedRelative.md5() }
+
+    val sourceFiles = generated.sourceFiles
+      .map { AbsoluteFile(it).relativeTo(projectDir) }
+      .ifEmpty { listOf(RelativeFile.ANY) }
+
+    for (source in sourceFiles) {
+
+      val hasCycle = getGeneratedFilesRecursive(generatedRelative).any { it == source }
+
+      if (hasCycle) {
+        throw AnvilCompilationException(
+          """
+          Adding this mapping would create a cycle.
+             source path: ${source.absolute(projectDir).file}
+          generated path: ${generated.file}
+          """.trimIndent(),
+        )
+      }
+
+      sourcesToGenerated.add(source, generatedRelative)
+      generatedToSources.add(generatedRelative, source)
+
+      // At this point, any source file should be in the md5 map already.
+      // But in case it isn't, we'll add it now.
+      addMd5(source, overwrite = false)
+    }
+  }
+
+  fun removeSource(sourceFile: FileType) {
+    when (sourceFile) {
+      is AbsoluteFile -> removeSource(sourceFile.relativeTo(projectDir))
+      is RelativeFile -> {
+        filesToMd5.remove(sourceFile)
+        generatedToSources.remove(sourceFile)
+
+        // All generated files that claim this source file as a source.
+        val generated = sourcesToGenerated.remove(sourceFile) ?: return
+
+        for (gen in generated) {
+
+          // For any generated file that has multiple sources,
+          // remove that generated file from the "generated" set for those other sources.
+          // Note that this does not call `removeSource` for that other source.
+          for (otherSource in generatedToSources[gen]) {
+            if (otherSource != sourceFile) {
+              sourcesToGenerated.remove(otherSource, gen)
+            }
+          }
+          removeSource(gen)
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns `true` if the [sourceFile] has changed since the last time it was added to the cache.
+   */
+  fun hasChanged(sourceFile: FileType): Boolean {
+    return when (sourceFile) {
+      is AbsoluteFile -> hasChanged(sourceFile.relativeTo(projectDir))
+      is RelativeFile -> {
+        if (sourceFile == RelativeFile.ANY) return true
+        val currentMd5 = sourceFile.md5()
+        val previousMd5 = filesToMd5.put(sourceFile, currentMd5)
```

Comment:
This should be using `get()` right? Right now it has a side effect of modifying the tracked MD5

Reply:
This is intentional, but it's tricky.

This function must always calculate the file's current md5 in order to do the comparison.  As such, we might as well cache it.  The `filesToMd5.put(...)` returns the previously-stored value or null, so we still get to do the same comparison -- and now we get the cache update for free.

Reply:
👍 that makes sense for optimizing. I don't remember if there's already a test for this, but if not it's probably worth adding one to make it explicit that this behavior is intentional. I would have assumed it was a bug and changed it to `get` if I was reading this sometime in the future since the method name `hasChanged` suggests it's only checking things